source/client_channel_handler.c

```diff
@@ -602,8 +626,42 @@ static void s_request_timeout_task(struct aws_channel_task *task, void *arg, enu
             "%" PRIu16 ".",
             (void *)task,
             request->message_id);
-        /* If task cancelled, assume safe shutdown is in progress and signal hash_table destroy to clean up */
-        request->cancelled = true;
+        /*
+         * If the task is cancelled, assume safe shutdown is in progress.
+         * There are two distinct situations where requests end up with a cancelled task:
+         *   (1) The request has been completed but disconnect was begun before the task had a chance to run
+         *       normally.  We need a way to ensure that aws_memory_pool_release is called for this request.
+         *
+         *       We can do that by
+         *          (1) setting request->cancelled to true.  Then the hash table's value removal function will
+         *              ensure aws_memory_pool_release is called when the table is cleared during final disconnect.
+         *          (2) If we're uncomfortable (semantically) with a request that is both "completed" and "cancelled"
+         *              then we could alternatively call aws_memory_pool_release here.
+         *
+         *       If we do not do one of these two, the request will leak and the pool will not clean up properly.
+         *
+         *   (2) The request is incomplete.  In this case, we need to call complete with an error and also
+         *       ensure that the hash table removal function calls the memory pool release, by setting
+         *       cancelled to true.
+         */
+        if (!request->cancelled) {
+            request->cancelled = true;
+            if (!request->completed) {
+                AWS_LOGF_DEBUG(
+                    AWS_LS_MQTT_CLIENT,
+                    "static: task id %p, completing with interrupt request for packet id "
+                    "%" PRIu16 ".",
+                    (void *)task,
+                    request->message_id);
+                if (request->on_complete) {
+                    request->on_complete(
```

Comment:
should we be setting `request->completed = true`?
since we're firing the on_complete callback?

Reply:
True, task shouldn't get rerun but its safer to be sure.